hooks/useConsultants.js

```diff
@@ -0,0 +1,93 @@
+import { useEffect, useState } from 'react';
+import { fetchUsers } from '../services/searchUser';
+
+const useConsultants = (
+	searchQuery,
+	selectedSpecialities,
+	experience,
+	selectedRegions,
+	price,
+) => {
+	const [filteredConsultants, setFilteredConsultants] = useState([]);
+	const [isLoading, setIsLoading] = useState(true);
+	const [fetchError, setFetchError] = useState(null);
+
+	useEffect(() => {
+		let timeoutId;
+
+		const fetchData = async () => {
+			try {
+				const data = await fetchUsers();
+				const filteredData = filterConsultants(data, searchQuery);
+				setFilteredConsultants(filteredData);
+				setIsLoading(false);
+			} catch (error) {
+				setFetchError(error);
+				setIsLoading(false);
+			}
+		};
+
+		// Use a timeout to debounce the fetchData function
+		const debounceFetchData = () => {
+			clearTimeout(timeoutId);
+			setIsLoading(true);
+			timeoutId = setTimeout(fetchData, 600); // Adjust the delay as needed
+		};
+
+		debounceFetchData();
+
+		return () => {
+			clearTimeout(timeoutId); // Clear the timeout on component unmount
+		};
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, [searchQuery, selectedSpecialities, experience, selectedRegions, price]);
+
+	const filterConsultants = (data) => {
```

Comment:
1. Notice that this function is a stateless function, which means it doesn't have to be aware of the state/context of any component if it is passed enough arguments. Since it can be turned into a stateless function, it should be turned into a stateless helper/util function (positioned in the `util` folder) as it will help later with the debugging. Note that this comment only applies as long as we __do the filtering on the frontend__

2. In the future, we might want to filter the data using Firebase (Firebase has functions that directly support filtering, we don't even need Cloud Functions) or a dedicated backend, this function will be refactored to make a call to a server. Then, it shall be put into the `service` folder.

Reply:
The stateless function being put in the util folder makes sense. 
Also im glad you brought up how we are going to be filtering. 
These are my thoughts and would be glad to hear what you have to say. 
I carefully considered the approach for filtering consultants in the hook, anticipating that Firebase functions might handle the filtering. I weighed the following considerations:

Performance: Should we fetch all consultants and filter on the frontend, or make multiple calls based on different filters?
Number of Calls: Instead of making individual calls for each filter (e.g., each button click), would it be more efficient?
Stacking Filters: If we opt for individual calls, how do we handle stacking filters?
For the initial implementation, I chose a single call with filtering on the frontend for simplicity in prototyping. I recognize that this may change if we find that utilizing Firebase functions for filtering is more practical in the long run.
Let me know your thoughts!

Reply:
The solution to filtering performance doesn't have to be one or another, it can be both frontend and backend filtering, and even more. Heck, we can start throwing browser and server caching in if we are really crazy about optimization.

However, back to our modest app, I think we should test out both approaches, and see which one is more performant. I suspect that letting Firebase do the filtering will be more efficient, because Firebase indexes each field in the document, so the search is faster than scanning the entire dataset (which is normally how we implement frontend filtering in simple apps like this). Of course, the constant communication between our frontend and Firebase will take some time (and therefore I suggested we should test both approaches), but I doubt that our users will use the filters too often, so the total communication waiting time won't be that significant.

Reply:
Yea sure we can test this out as soon as the querys are up and running and the database has some dummy data.